### PR TITLE
Instrumenting switch cases to detect taint that influence control flow

### PR DESCRIPTION
**Pull request by hbrodin**
Previously, only branch instructions was only considered to affect control flow. This PR instruments switch instructions as well. Switch instructions do also affect control flow.

### CODE REVIEW

File: polytracker/src/passes/polytracker_pass.cpp [hunk 1]
```diff
@@ -132,6 +135,14 @@ void PolytrackerPass::visitBranchInst(llvm::BranchInst &BI) {
   }
 }
 
+void PolytrackerPass::visitSwitchInst(llvm::SwitchInst &SI) {
+  // TODO (hbrodin): Is a guard needed? E.g. op_check as in visitBranchInst?
+  llvm::IRBuilder<> IRB(&SI);
+  llvm::LLVMContext &context = mod->getContext();
```
Comment (surovic) [inline polytracker/src/passes/polytracker_pass.cpp:141]: This seems redundant?

File: polytracker/src/passes/polytracker_pass.cpp [hunk 2]
```diff
@@ -589,6 +600,8 @@ bool PolytrackerPass::runOnModule(llvm::Module &mod) {
         for (auto &inst : bb) {
           if (auto *BI = llvm::dyn_cast<llvm::BranchInst>(&inst)) {
             visitBranchInst(*BI);
+          } else if (auto *SI = llvm::dyn_cast<llvm::SwitchInst>(&inst)) {
```
Comment (surovic) [inline polytracker/src/passes/polytracker_pass.cpp:603]: We can avoid the whole `for(auto &inst : bb){...}` by using `visit(bb->getTerminator())`.

**Review by surovic — APPROVED** [review]

File: polytracker/src/passes/polytracker_pass.cpp (hunk 2 above)
Comment (hbrodin) [reply polytracker/src/passes/polytracker_pass.cpp:603]: Great idea! We can't do exactly as you write, but we don't need to loop. If we use the visitor we would also get e.g. the ret-instr visitation and that is not desired.

File: polytracker/src/passes/polytracker_pass.cpp (hunk 1 above)
Comment (hbrodin) [reply polytracker/src/passes/polytracker_pass.cpp:141]: True

File: polytracker/src/passes/polytracker_pass.cpp (hunk 2 above)
Comment (surovic) [reply polytracker/src/passes/polytracker_pass.cpp:603]: True, my bad! Does it matter when we visit branches and switches?

Comment (hbrodin) [reply polytracker/src/passes/polytracker_pass.cpp:603]: We would add instrumentation for the ret instruction and that typically instruments the instruction with a call to a method to log function exits.  I'll make a slight variation.

**Pull request merged**